consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftBlockHeightManager.java

```diff
@@ -12,35 +12,217 @@
 
-  public void handleRoundChangeMessage(final SignedData<RoundChangePayload> roundChangeMsg) {}
+    // Its possible the locally created RoundChange triggers the transmission of a NewRound
+    // message - so it must be handled accordingly.
+    handleRoundChangeMessage(localRoundChange);
+  }
+
+  public void handleProposalMessage(final SignedData<ProposalPayload> msg) {
+    LOG.info("Received a Proposal message.");
+    actionOrBufferMessage(
+        msg,
+        () -> currentRound.handleProposalMessage(msg),
+        (roundstate) -> roundstate.setProposedBlock(msg));
+  }
+
+  public void handlePrepareMessage(final SignedData<PreparePayload> msg) {
+    LOG.info("Received a prepare message.");
+    actionOrBufferMessage(
+        msg,
+        () -> currentRound.handlePrepareMessage(msg),
+        (roundstate) -> roundstate.addPrepareMessage(msg));
+  }
+
+  public void handleCommitMessage(final SignedData<CommitPayload> msg) {
+    LOG.info("Received a commit message.");
+    actionOrBufferMessage(
+        msg,
+        () -> currentRound.handleCommitMessage(msg),
+        (roundstate) -> roundstate.addCommitMessage(msg));
+  }
+
+  private void actionOrBufferMessage(
+      final SignedData<? extends Payload> msg,
+      final Action inRoundHandler,
+      final Consumer<RoundState> buffer) {
+    final ConsensusRoundIdentifier msgRoundId = msg.getPayload().getRoundIdentifier();
+    if (msgRoundId.equals(currentRound.getRoundIdentifier())) {
+      inRoundHandler.apply();
+    } else if (msgRoundId.getRoundNumber() > currentRound.getRoundIdentifier().getRoundNumber()) {
+      final RoundState roundstate =
+          futureRoundStateBuffer.computeIfAbsent(
+              msgRoundId.getRoundNumber(), k -> roundStateCreator.apply(msgRoundId));
+      buffer.accept(roundstate);
+    }
+  }
+
+  public void handleRoundChangeMessage(final SignedData<RoundChangePayload> msg) {
+    final Optional<RoundChangeCertificate> result =
+        roundChangeManager.appendRoundChangeMessage(msg);
+
+    if (result.isPresent()) {
+      final ConsensusRoundIdentifier targetRound = msg.getPayload().getRoundIdentifier();
+      if (targetRound.getRoundNumber() > currentRound.getRoundIdentifier().getRoundNumber()) {
+        startNewRound(targetRound.getRoundNumber());
+      }
+
+      if (finalState.isLocalNodeProposerForRound(targetRound)) {
+        currentRound.startRoundWith(result.get(), clock.millis() / 1000);
+      }
+    }
+  }
 
-  public void handleNewRoundMessage(final SignedData<NewRoundPayload> newRoundMsg) {}
+  private void startNewRound(final int roundNumber) {
+    if (futureRoundStateBuffer.containsKey(roundNumber)) {
+      currentRound =
+          roundFactory.createNewRoundWithState(
+              parentHeader, futureRoundStateBuffer.get(roundNumber));
+      futureRoundStateBuffer.entrySet().removeIf(k -> k.getKey() <= roundNumber);
```

Comment:
nit: if you use the futureRoundStateBuffer.keySet you can slightly simply the lambda as you won't need to call getKey

Reply:
done.